packages/abi-coder/src/types/JsonAbiNew.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Types for Fuel JSON ABI Format as defined on:
+ * https://github.com/FuelLabs/fuel-specs/blob/master/src/abi/json-abi-format.md
+ */
+export interface JsonAbi {
+  readonly specVersion: string;
+  readonly encodingVersion: string;
+  readonly programType: string;
+  readonly concreteTypes: readonly ConcreteType[];
+  readonly metadataTypes: readonly MetadataType[];
+  readonly functions: readonly AbiFunction[];
+  readonly loggedTypes: readonly LoggedType[];
+  readonly messagesTypes: readonly MessageType[];
+  readonly configurables: readonly Configurable[];
+}
+
+export interface ConcreteType {
+  readonly type: string;
+  readonly concreteTypeId: string;
+  readonly metadataTypeId?: number;
+  readonly typeArguments?: readonly string[];
```

Comment:
Is this the same as the old `typeParameters`?

Reply:
@sdankel I suggest you check out the spec changes _([#596](https://github.com/FuelLabs/fuel-specs/pull/596), [#599](https://github.com/FuelLabs/fuel-specs/pull/599))_ to get an understanding of the new format.

To answer the question, no it's not because `typeParameters` contained a list of used generic types whereas the `typeArguments` here are a list of concrete types that will be replacing the type parameters of the linked metadata type.

Reply:
Okay, so is this an array of concrete type hashes? It would be much easier to use if the SDK didn't expose the hashes at all and was consistent about using array indexes everywhere.

It would be nice also if the SDK would decode the hashes.

Reply:
The SDK just consumes the ABI as per the specification and doesn't modify it. Concrete types and their hashed ids are now the way to identify actual directly used types (input or output, log, configurable, etc.) in a sway program. You can decode the hashes via the `Interface.decodeType(concreteTypeId: string, data: Uint8Array)` method which we added in this PR to enable users to decode any concrete type.

Reply:
Thanks for the pointer to the helper function. Does it also work for decoding metadata type hashes?

Reply:
Hashes are reserved exclusively to concrete types. Metadata types are _never_ used by the users directly so the SDK doesn't provide methods to work with them. I believe that you also won't have a need for them in sway-playground in relation to encoding/decoding as function inputs/outputs are always concrete types.